pkg/build/generator/generator.go

```diff
@@ -258,24 +288,57 @@ func (g *BuildGenerator) resolveImageRepoReference(ctx kapi.Context, from *kapi.
 	} else {
 		namespace = defaultNamespace
 	}
-	imageStream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), from.Name)
-	if err != nil {
-		if errors.IsNotFound(err) {
+
+	switch from.Kind {
+	case "ImageStreamImage":
+		bits := strings.Split(from.Name, "@")
+		stream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), bits[0])
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return "", err
+			}
+			return "", fatalError{err}
+		}
+		ref, err := imageapi.DockerImageReferenceForStream(stream)
+		if err != nil {
 			return "", err
 		}
-		return "", fatalError{err}
-	}
-
-	if imageStream == nil || len(imageStream.Status.DockerImageRepository) == 0 {
-		return "", fmt.Errorf("could not resolve image stream %s/%s", namespace, from.Name)
-	}
-	glog.V(4).Infof("Found image stream %s/%s", namespace, imageStream.Name)
+		ref.ID = bits[1]
+		return ref.String(), nil
+	case "ImageStreamTag":
+		bits := strings.Split(from.Name, ":")
+		stream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), bits[0])
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return "", err
+			}
+			return "", fatalError{err}
+		}
+		latest, err := imageapi.LatestTaggedImage(stream, bits[1])
+		if err != nil {
+			return "", err
+		}
+		return latest.DockerImageReference, nil
+	default:
+		// assumes ImageStream for historic purposes
+		imageStream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), from.Name)
```

Comment:
since tag could be "" here, i'd rather defer the decision about what to default it to into the call to LatestTaggedImage.  (at least it's one less place where i'm hardcoding "latest")